apstools/beamtime/bss_info.py

```diff
@@ -0,0 +1,459 @@
+
+
+def trim(text, length=40):
+    if len(text) > length:
+        text = text[:length-3] + "..."
+    return text
+
+
+class DmRecordNotFound(Exception): ...
+class EsafNotFound(DmRecordNotFound): ...
+class ProposalNotFound(DmRecordNotFound): ...
+
+
+class EsafInfo:
+    
+    def __init__(self, esaf):
+        self.title = esaf["esafTitle"]
+        self.id = esaf["esafId"]
+        self.status = esaf["esafStatus"]
+        self.startDate = esaf["experimentStartDate"]
+        self.endDate = esaf["experimentEndDate"]
+        self.sector = esaf["sector"]
+        users = [u["lastName"] for u in esaf["experimentUsers"]]
+        self.users = ",".join(sorted(users))
+
+
+class ScheduleInfo:
+    
+    def __init__(self, proposal):
+        self.others = []
+        self.contact = None
+        self.title = proposal["title"]
+        self.id = proposal["id"]
+        for exp in proposal["experimenters"]:
+            if exp.get("piFlag", "N") == "Y":
+                self.badge = exp["badge"]
+                self.email = exp["email"]
+                self.institution = exp["institution"]
+                self.contact = f'{exp["firstName"]} {exp["lastName"]}'
+            else:
+                self.others.append(exp["lastName"])
+        self.others = ",".join(sorted(self.others))
+
+
+def get_options():
+    parser = argparse.ArgumentParser(
+        prog=os.path.split(sys.argv[0])[-1], 
+        description=__doc__.strip().splitlines()[0],
+        )
+
+    subcommand = parser.add_subparsers(dest='subcommand', title='subcommand')
+
+    p_sub = subcommand.add_parser('beamlines', help="print list of beamlines")
+
+    p_sub = subcommand.add_parser('current', help="print current ESAF(s) and proposal(s)")
+    p_sub.add_argument('beamlineName', type=str, help="Beamline name")
+
+    p_sub = subcommand.add_parser('cycles', help="print APS cycle names")
+
+    p_sub = subcommand.add_parser('esaf', help="print specific ESAF")
+    p_sub.add_argument('esafId', type=int, help="ESAF ID number")
+
+    p_sub = subcommand.add_parser('proposal', help="print specific proposal")
+    p_sub.add_argument('proposalId', type=str, help="ESAF ID number")
+    p_sub.add_argument('cycle', type=str, help="APS run (cycle) name")
+    p_sub.add_argument('beamlineName', type=str, help="Beamline name")
+
+    p_sub = subcommand.add_parser('clear', help="EPICS PVs: clear")
+    p_sub.add_argument('prefix', type=str, help="EPICS PV prefix")
+
+    p_sub = subcommand.add_parser('setup', help="EPICS PVs: setup")
+    p_sub.add_argument('prefix', type=str, help="EPICS PV prefix")
+    p_sub.add_argument('beamlineName', type=str, help="Beamline name")
+    p_sub.add_argument('cycle', type=str, help="APS run (cycle) name")
+
+    p_sub = subcommand.add_parser('update', help="EPICS PVs: update from BSS")
+    p_sub.add_argument('prefix', type=str, help="EPICS PV prefix")
+
+    return parser.parse_args()
+
+
+def main():
```

Comment:
Codacy found an issue: [main is too complex (19) (MC0001)](https://app.codacy.com/gh/BCDA-APS/apstools/pullRequest?prid=5764884)